### ISSUE TITLE
Invalid missing_return warning

### ISSUE THREAD

**Issue by kinex** [dart-lang/linter#1645]
This sample code results incorrectly missing_return warning:
```dart
  bool test() {
    [1, 2].forEach((item) {
      if (item == 1) {
        return false;
      }
    });
    return true;
  }
```
Flutter 1.7.8+hotfix.3 • channel stable • https://github.com/flutter/flutter.git
Framework • revision b712a172f9 (9 hours ago) • 2019-07-09 13:14:38 -0700
Engine • revision 54ad777fd2
Tools • Dart 2.4.0

**Comment by a14n** [comment]
The lint correctly flag that your lambda inside `forEach` doesn't return value if `item != 1`.

`return false` in this lambda provides `false` as result of the lambda and not as result of `test` (dart does not support non-local return). If you want this false value to be returned as result of `test` you must not use `forEach`. You need to use a for-loop

```dart
  bool test() {
    for (var item in [1, 2]) {
      if (item == 1) {
        return false;
      }
    }
    return true;
  }
```

**Comment by kinex** [comment]
Yes you are right, this is actually a stupid bug in my production code. But since the signature of the lambda for ```foreach``` is  ```void f(E element)```, I would expect the warning be something like "this function should not return a value".

**Comment by srawlins** [comment]
@kinex You make a really good point. The issue you're hitting is that `void` is considered one of the Top types (with `Object` and `dynamic`). In some aspects of the type system, the top types are considered equivalent to each other. So this is statically sound code:

```dart
void fn() => 1==2 ? false : "hello";
```

So, awkwardly, is this:

```dart
[1,2,3].forEach((e) {
  if (1==1) return true;
  else return 7;
});
```

This is because Dart looks at the Least Upper Bound between the various types (bool and String in the first example, bool and int in the second), and comes up with Object. Object is sort of close enough to void, so there are no static checks that report "Object is not void."

The `missing_return` error in this case is basically pointing out that your callback _sometimes_ returns a value, and sometimes it doesn't. That's the smell it would like to point out.

**Comment by kinex** [comment]
Thanks for the details. I think current `missing_return `error is better than no error at all also in this `foreach` case. At least it indicates that something is wrong. 

Currently I would not get any error for this:
```Dart
bool test() {
  [1, 2].forEach((item) {
    if (item == 1) {
      return false;
    } else {
      return true;
    }
  });
  return true;
}
```

But this results error `return_of_invalid_type` :
```Dart
void test2() {
  return true;
}
```
I would like to see the error `return_of_invalid_type` also for the `foreach` case, but I understand it may not be technically possible or at least not very easy.
